### PR TITLE
Add pagination support for Sentry endpoints

### PR DESCRIPTION
**Pull request by olegpixel**
### Overview

This PR introduces pagination support, enhancing our ability to retrieve data from Sentry's API efficiently.

Sentry Projects Endpoint - https://docs.sentry.io/api/organizations/list-an-organizations-projects/
Sentry Pagination Docs - https://docs.sentry.io/api/pagination/
Link Header - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link

### Changes Made

Added a new `FetchWithPagination` method in the SentryClient to parse the Link header and extract the next page URL when available, aligning our code with Sentry's pagination guidelines.

Added a new `GetAllProjects` function to utilise the enhanced Fetch method, enabling it to fetch projects across multiple pages as needed.

### CODE REVIEW

File: pkg/plugin/handlers_resources.go [hunk 1]
```diff
@@ -33,7 +33,7 @@ func GetProjectsHandler(client *sentry.SentryClient) http.HandlerFunc {
 			http.Error(rw, "invalid orgSlug", http.StatusBadRequest)
 			return
 		}
-		orgs, err := client.GetProjects(orgSlug)
+		orgs, err := client.GetAllProjects(orgSlug)
```
Comment (taj-p) [inline pkg/plugin/handlers_resources.go]: I think it's a little confusing that the `GetProjectsHandler` doesn't use `GetProjects`

File: pkg/sentry/projects.go [hunk 2]
```diff
@@ -26,6 +26,7 @@ type SentryProject struct {
 	} `json:"teams"`
 }
 
+// fetch first 100 projects
 func (sc *SentryClient) GetProjects(organizationSlug string) ([]SentryProject, error) {
```
Comment (taj-p) [inline pkg/sentry/projects.go]: Are there cases where we would ever want 2 `Get*Project` APIs one for first 100 and other for all? 

Perhaps this function could take an argument that limits pagination or similar and just have 1 `GetProjects` function. I see that the only other reference is to the health endpoint. Perhaps the health endpoint just fetches the first page, but the `GetProjects` handler returns all pages?

Probably simpler is to replace `GetProjects` with your `GetAllProjects` implementation and use the `GetOrganizations` function for the healthcheck since that isn't paginated and shouldn't ever need to be?

File: pkg/sentry/projects.go [hunk 3]
```diff
@@ -35,6 +36,25 @@ func (sc *SentryClient) GetProjects(organizationSlug string) ([]SentryProject, e
...
 }
 
+// Fetch all projects, checking if a link for the next page is specified in the headers
+func (sc *SentryClient) GetAllProjects(organizationSlug string) ([]SentryProject, error) {
+	allProjects := []SentryProject{}
+	url := "/api/0/organizations/" + organizationSlug + "/projects/"
+
+	for (url != "") {
+		projects := []SentryProject{}
+		nextURL, err := sc.FetchWithPagination(url, &projects)
```
Comment (taj-p) [inline pkg/sentry/projects.go]: It doesn't look like this API handles rate limits - should we wrap this in an exponential backoff? Perhaps we should do this in a separate PR for all endpoints another time? 😄

**Review by taj-p — APPROVED** [review]
LGTM! Just a comment on `GetAllProjects` vs `GetProjects`. Looking forward to seeing how the Sentry folk review the PR

File: pkg/sentry/projects.go (hunk 3 above)
Comment (foadnh) [reply pkg/sentry/projects.go]: :+1: Maybe you can wrap request sending logic in a separate package and use a decorator pattern to add rate limits check, backoff and retry logics?
I agree it is better to be a separate PR.

File: pkg/sentry/sentry.go [hunk 4]
```diff
@@ -8,6 +8,7 @@ import (
 	"strings"
 
 	"github.com/grafana/grafana-plugin-sdk-go/build"
+	"github.com/peterhellberg/link"
```
Comment (foadnh) [inline pkg/sentry/sentry.go:11]: This repo only has 40 starts, and it doesn't seem well-maintained! There was only one commit in the last year.
I suggest not using it.

File: pkg/sentry/sentry.go [hunk 5]
```diff
@@ -34,6 +35,44 @@ type SentryErrorResponse struct {
...
+			return "", err
+		}
+	} else {
+		var errResponse SentryErrorResponse
+		if err := json.NewDecoder(res.Body).Decode(&errResponse); err != nil {
+			errorMessage := strings.TrimSpace(fmt.Sprintf("%s %s", res.Status, err.Error()))
+			return "", errors.New(errorMessage)
+		}
+		errorMessage := strings.TrimSpace(fmt.Sprintf("%s %s", res.Status, errResponse.Detail))
+		return "", errors.New(errorMessage)
```
Comment (foadnh) [inline pkg/sentry/sentry.go:71]: It would be nice to move this part to
```
func (err SentryErrorResponse) Error() string {
     // ...
}
```
and in here simply `return "", errResponse`

**Review by foadnh — CHANGES_REQUESTED** [review]

**Review by foadnh — APPROVED** [review]

File: pkg/sentry/projects.go (hunk 3 above)
Comment (olegpixel) [reply pkg/sentry/projects.go]: I'm planning to create a separate PR for the rate limiting handling. We have `sentryHttpClient`, so I'm planing to put the login there.

File: pkg/sentry/sentry.go (hunk 4 above)
Comment (olegpixel) [reply pkg/sentry/sentry.go:11]: This is the most popular repo I found. I think I can implement the method I need on my own. Do you think it worth it?

Comment (foadnh) [reply pkg/sentry/sentry.go:11]: Yes, I think it is a good idea to implement your own method. Thanks Oleg.

Comment (olegpixel) [reply pkg/sentry/sentry.go:11]: I've checked the repo once again, and it seems that the entire code consists of just a 'Parse' method. So, if I decide to proceed with my implementation, I'll likely end up with a similar function. If you don't mind, I might leave this for further consideration later.

File: pkg/sentry/projects.go (hunk 2 above)
Comment (olegpixel) [reply pkg/sentry/projects.go]: I've updated the 'GetProjects' endpoint by merging it with 'GetAllProjects' and adding an additional flag to determine whether it should request all projects or just one page.

File: pkg/sentry/sentry.go (hunk 5 above)
Comment (olegpixel) [reply pkg/sentry/sentry.go:71]: I'm not sure I understand what you mean. Could you please provide more details?

File: pkg/sentry/sentry.go (hunk 4 above)
Comment (foadnh) [reply pkg/sentry/sentry.go:11]: Yes, that is fine. We can do it later.

File: pkg/sentry/sentry.go (hunk 5 above)
Comment (foadnh) [reply pkg/sentry/sentry.go:71]: I wanted to suggest something like:

```Go
type SentryErrorResponse struct {
	Detail string `json:"detail"`
	decodingErr error
	responseStatus string
}

func (err *SentryErrorResponse) Error() string {
	if err.decodingErr != nil {
		return strings.TrimSpace(fmt.Sprintf("%s %s", res.Status, err.Error()))
	}
	return strings.TrimSpace(fmt.Sprintf("%s %s", res.Status, errResponse.Detail))
}

func responseError(res *http.Response) error {
	err := &SentryErrorResponse{
		responseStatus res.Status
	}
	err.decodingErr = json.NewDecoder(res.Body).Decode(err)
	return err
}

func (sc *SentryClient) FetchWithPagination(path string, out interface{}) (string, error) {
	// ...
	if res.StatusCode == http.StatusOK {
		// ...
	} else {
		return "", responseError(res)
	}
```

But then I realised `SentryErrorResponse` is also used in the `Fetch()` method. So ignore me :-)
